theano/gof/tests/test_vm.py

```diff
@@ -369,3 +369,93 @@ def check_storage(storage_map):
         assert check_storage(storage_map)[0]
         assert len(set([id(v) for v in
                         storage_map.values()])) < len(storage_map)
+
+
+def test_shape_reallocation():
+        a, b, c, d = [tensor.dvector(n) for n in ['a', 'b', 'c', 'd']]
+        # No Shape Infer
+        z = 3*a + b
+        x = c + 4*d
+        y = z + x
+        # With Shape Infer
+        o = (3*a + b)
+        p = (c + 4*d)
+        w = o + p
```

Comment:
Why there should be a difference between y and w? They are equivalent graph without any diff. This should not change anything for the infer_shape. You check the same things on both.